### ISSUE TITLE
`cargo doc` fails for the h2 crate (v0.1.9) on nightly 2018-06-05

### ISSUE THREAD

**Issue by jimmycuadra** [rust-lang/rust#51468]
```
$ rustup show
Default host: x86_64-apple-darwin

installed toolchains
--------------------

stable-x86_64-apple-darwin
nightly-x86_64-apple-darwin (default)

active toolchain
----------------

nightly-x86_64-apple-darwin (default)
rustc 1.28.0-nightly (4a9c58c6b 2018-06-05)
$ cargo doc -p h2
 Documenting h2 v0.1.9
warning: [u8] cannot be resolved, ignoring it...
  --> /Users/example/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-0.4.8/src/lib.rs:1:1
   |
1  | / //! Provides abstractions for working with bytes.
2  | | //!
3  | | //! The `bytes` crate provides an efficient byte buffer structure
4  | | //! ([`Bytes`](struct.Bytes.html)) and traits for working with buffer
...  |
68 | | //! perform a syscall, which has the potential of failing. Operations on `Buf`
69 | | //! and `BufMut` are infallible.
   | |________________________________^
   |
   = note: the link appears in this line:

            A `Bytes` handle can be created directly from an existing byte store (such as &[u8]
                                                                                            ^^

warning: [cfg] cannot be resolved, ignoring it...
  --> /Users/example/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.3/src/lib.rs:1:1
   |
1  | / //! A macro for defining #[cfg] if-else statements.
2  | | //!
3  | | //! The macro provided by this crate, `cfg_if`, is similar to the `if/elif` C
4  | | //! preprocessor macro by allowing definition of a cascade of `#[cfg]` cases,
...  |
26 | | //! # fn main() {}
27 | | //! ```
   | |_______^
   |
   = note: the link appears in this line:

            A macro for defining #[cfg] if-else statements.
                                   ^^^

warning: [cfg] cannot be resolved, ignoring it...
  --> /Users/example/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.3/src/lib.rs:1:1
   |
1  | / //! A macro for defining #[cfg] if-else statements.
2  | | //!
3  | | //! The macro provided by this crate, `cfg_if`, is similar to the `if/elif` C
4  | | //! preprocessor macro by allowing definition of a cascade of `#[cfg]` cases,
...  |
26 | | //! # fn main() {}
27 | | //! ```
   | |_______^
   |
   = note: the link appears in this line:

            This allows you to conveniently provide a long list #[cfg]'d blocks of code
                                                                  ^^^

warning: [Debug] cannot be resolved, ignoring it...
   |
   = note: the link appears in this line:

            themselves through the [`Display`] and [`Debug`] traits, and may provide
                                                    ^^^^^^^

error[E0658]: access to extern crates through prelude is experimental (see issue #44660)
  |
  = help: add #![feature(extern_prelude)] to the crate attributes to enable

error: Could not document `h2`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name h2 /Users/example/.cargo/registry/src/github.com-1ecc6299db9ec823/h2-0.1.9/src/lib.rs -o /Users/example/example-cargo-project/target/doc -L dependency=/Users/example/example-cargo-project/target/debug/deps --extern byteorder=/Users/example/example-cargo-project/target/debug/deps/libbyteorder-df3eee4befa1312c.rmeta --extern bytes=/Users/example/example-cargo-project/target/debug/deps/libbytes-9fc866f4058b4096.rmeta --extern fnv=/Users/example/example-cargo-project/target/debug/deps/libfnv-2278a5d27eb0baf9.rmeta --extern futures=/Users/example/example-cargo-project/target/debug/deps/libfutures-5c1e38662d6f78e5.rmeta --extern http=/Users/example/example-cargo-project/target/debug/deps/libhttp-b2084a070c91801e.rmeta --extern indexmap=/Users/example/example-cargo-project/target/debug/deps/libindexmap-fbca607fdf3e2794.rmeta --extern log=/Users/example/example-cargo-project/target/debug/deps/liblog-1c027577d3857644.rmeta --extern slab=/Users/example/example-cargo-project/target/debug/deps/libslab-4299e706602486cd.rmeta --extern string=/Users/example/example-cargo-project/target/debug/deps/libstring-1c5b00744b69ff86.rmeta --extern tokio_io=/Users/example/example-cargo-project/target/debug/deps/libtokio_io-79182f96ea980954.rmeta` (exit code: 101)
```

**Comment by pietroalbini** [comment]
This is a regression from stable to nightly. cc @rust-lang/rustdoc

**Comment by GuillaumeGomez** [comment]
No it's not, it's a nightly version.

**Comment by QuietMisdreavus** [comment]
We've been getting enough of these breaking errors out of the intra-links that i wonder if we should introduce a "rustdoc feature flag" for it. Something like `#![doc(feature(type_links))]`? It would help contain these errors.

The actual compile error is also a duplicate of https://github.com/rust-lang/rust/issues/50561. Everything else is just warnings that lead up to that point.`

**Comment by cramertj** [comment]
@QuietMisdreavus Is there an issue somewhere tracking the the resolution of interactions between `#![deny(warnings]` and rustdoc? Similar to how `cargo build` works, `cargo doc` shouldn't fail if upstream dependencies have `#![deny(warnings)]`-- I'd even argue that denying warnings shouldn't ever cause rustdoc to fail, and that there should be a separate flag for ensuring that documentation doesn't trigger warnings. Do you know if there's been any talk about a "please, really do document these crates to the best of your ability" setting in `rustdoc`?

**Comment by GuillaumeGomez** [comment]
Why should rustdoc overlook lints/warnings/errors? It seems to be problematic from my point of view. Thanks to it, we were able to find errors while documenting the std so I'd say it's a great thing which was missing.

Otherwise I opened a PR to fix the `#[deny(warnings)]` issue in #51956. Waiting for the compiler team to end discussions about it.

**Comment by cramertj** [comment]
@GuillaumeGomez It's a great thing that rustdoc can issue warnings, but it shouldn't block the creation of docs. That is, for an upstream with `#![deny(warnings)]`, there is *no way* to generate documentation for it when a new warning is added and the doc generation breaks.

**Comment by GuillaumeGomez** [comment]
@cramertj: If it blocks compilation, it should block doc generation.

@QuietMisdreavus: Yes and we need to fix a few things but that's pretty common for new features.

**Comment by rtsuk** [comment]
@GuillaumeGomez In my case, it does not block compilation. I can build with my transitive dependency on futures_core just fine, it's only when I try to document my crate with its dependencies that cargo doc fails when trying to document futures_core. 

cargo doc worked as of a few weeks ago and is crucially important to me. I would really appreciate it if some solution were found to let it work again.

**Comment by QuietMisdreavus** [comment]
I wonder, now that https://github.com/rust-lang/rust/pull/51732 is landed, whether we can make cargo cap lints to Allow for dependencies when documenting, like it does for compiling? That would sidestep errors like these where the crate really isn't at fault. I just noticed those lint flags were unstable, though...

**Comment by GuillaumeGomez** [comment]
Once https://github.com/rust-lang/rust/pull/51956 is merged, your problem should be solved.

**Comment by dgrnbrg** [comment]
I am also running into this issue: I upgraded my toolchain to the latest nightly, and suddenly something in a dependency of a dependency of a dependency has now caused my doc generation to break. I will retry tomorrow when the next nightly is published.

**Comment by rtsuk** [comment]
Still not working with 2018-07-15, but I found that by setting RUSTDOCFLAGS to "--cap-lints allow -Z unstable-options" I can work around it.

**Comment by GuillaumeGomez** [comment]
Seems so! Closing it then.
